File: 101-symmetric-tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        # return self.is_mirror(root.left, root.right)
        return self.iterative(root)
        
    def is_mirror(self, left, right):
        if (left is None and right is None):
            return True
        if (left is None or right is None):
            return False
        return (left.val == right.val) and \
            self.is_mirror(left.left, right.right) and \
                self.is_mirror(left.right, right.left)
    
    def iterative(self, root):
        queue = [root, root]
        while len(queue) != 0:
            left = queue.pop(0)
            right = queue.pop(0)
            if (left is None and right is None):
                continue
            if (left is None or right is None) or (left.val != right.val):
                return False
            
            queue.append(left.left)
            queue.append(right.right)
            queue.append(left.right)
            queue.append(right.left)
        return True

File: 101-symmetric-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution1 {
public:
    bool isSymmetric(TreeNode* root) {
        return isMirror(root, root);
    }
    
    bool isMirror(TreeNode * left, TreeNode * right) {
        if ( !left && !right ) return true;
        if ( !left || !right ) return false;
        return ( left->val == right->val ) 
            && isMirror(left->left, right->right) 
            && isMirror(left->right, right->left);
    }
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        queue<TreeNode *> q;
        q.push(root); q.push(root);
        while ( !q.empty() ) {
            TreeNode * left = q.front(); q.pop();
            TreeNode * right = q.front(); q.pop();
            if ( !left && !right ) continue;
            if ( (!left || !right) || (left->val != right->val) ) return false;
            q.push(left->left);
            q.push(right->right);
            q.push(left->right);
            q.push(right->left);
        }
        return true;
    }
};